packages/functionals/botpress-qna/src/providers/microsoft_qna_maker.js

```diff
@@ -0,0 +1,171 @@
+    .filter(({ value }) => !_.isUndefined(value) && value !== '')
+    .value()
+
+export default class Storage {
+  constructor({ bp, config }) {
+    const baseURL = 'https://westus.api.cognitive.microsoft.com/qnamaker/v4.0'
+    const headers = { 'Ocp-Apim-Subscription-Key': config.microsoftQnaMakerApiKey }
+    Object.assign(this, { bp, client: axios.create({ baseURL, headers }) })
+  }
+
+  async initialize() {
+    const isBpKnowledgbase = ({ name }) => name === KNOWLEDGEBASE_NAME
+    const { data: { knowledgebases: initialKnowledgebases } } = await this.client.get('/knowledgebases/')
+    const existingKb = initialKnowledgebases.find(isBpKnowledgbase)
+    if (existingKb) {
+      this.knowledgebase = existingKb
+    } else {
+      const { data: { operationId } } = await this.client.post('/knowledgebases/create', { name: KNOWLEDGEBASE_NAME })
+      await this.waitForOperationToFinish(operationId)
+      const { data: { knowledgebases } } = await this.client.get('/knowledgebases/')
+      this.knowledgebase = knowledgebases.find(isBpKnowledgbase)
+    }
+
+    this.endpointKey = (await this.client.get('/endpointkeys')).data.primaryEndpointKey
+  }
+
+  publish = () => this.client.post(`/knowledgebases/${this.knowledgebase.id}`)
+
+  patchKb = params => this.client.patch(`/knowledgebases/${this.knowledgebase.id}`, params)
+
+  waitForOperationToFinish = async operationId => {
+    await Promise.delay(ms(200))
+    while (true) {
+      const { data, headers: { 'retry-after': timeout } } = await this.client.get(`/operations/${operationId}`)
+      if (!['Running', 'NotStarted'].includes(data.operationState)) {
+        return
+      }
+      this.bp.logger.info(
+        `[QNA] Waiting 3s for ${data.operationState} QnA Maker's #${operationId} operation to finish...`
+      )
+      await Promise.delay(ms(`${3} s`))
+    }
+  }
+
+  async update(data, id) {
+    const prevData = await this.getQuestion(id)
+
+    const questionsChanged = _.isEqual(data.questions, prevData.questions)
+    const questionsToAdd = _.difference(data.questions, prevData.questions)
+    const questionsToDelete = _.difference(prevData.questions, data.questions)
+
+    const { data: { operationId } } = await this.patchKb({
+      update: {
+        qnaList: [
+          {
+            id,
+            answer: data.answer,
+            ...(questionsChanged ? {} : { questions: { add: questionsToAdd, delete: questionsToDelete } }),
+            metadata: { delete: prevData.metadata, add: prepareMeta(data) }
+          }
+        ]
+      }
+    })
+
+    await this.waitForOperationToFinish(operationId)
+
+    this.invalidateCache()
+    await this.publish()
+    return id
+  }
+
+  async insert(qna) {
+    const qnas = _.isArray(qna) ? qna : [qna]
+    const { data: { operationId } } = await this.patchKb({
+      add: {
+        qnaList: qnas.map(qna => ({ id: 0, ..._.pick(qna, ['answer', 'questions']), metadata: prepareMeta(qna) }))
```

Comment:
Is `id: 0` required by their API and is it true that they actually assign proper IDs to it? Would make sense to leave a comment in this case

Reply:
It was in the example but it works without it either so I removed it.